--- a/audio/common/all-versions/default/service/service.cpp
+++ b/audio/common/all-versions/default/service/service.cpp
@@ -76,10 +76,10 @@
     signal(SIGPIPE, SIG_IGN);
 
     ::android::ProcessState::initWithDriver("/dev/vndbinder");
-    // start a threadpool for vndbinder interactions
-    ::android::ProcessState::self()->startThreadPool();
 
     ABinderProcess_setThreadPoolMaxThreadCount(1);
+    // start a threadpool for vndbinder interactions
+    ::android::ProcessState::self()->startThreadPool();
     ABinderProcess_startThreadPool();
 
     const int32_t defaultValue = -1;
